Kianoosh76/IBS
# [IBS-71] Refactored check submissions for student endpoint

## Description

- jira: https://ibs-utm.atlassian.net/browse/IBS-71
- Refactored check submission for student endpoint, and updated related files and models.

## Changes Made

- Added Submission model and migrations, which belong to the Group model, with the following columns:
    - `submission_id`: id of submission
    - `task`: id of the task, indicating which task is the submission for
    - `group_id`: group id of which the submission is for
    - `commit_id`: id of the commit history
    - `token_used`: number of token used for this submission
    - `timestamp`: time of submission, set to `DataTypes.NOW` by default
 

- Added token_count field to GroupUser which keeps tracking the number of tokens that are used for this submission.
<img width="195" alt="Screenshot 2024-04-25 at 2 44 25 PM" src="https://github.com/Kianoosh76/IBS/assets/71187784/88ddbcd7-ba5c-4459-a3fb-1696fe6d69e7">

- Updated behavior for the following helper function
    - `get_user_token_usage()`: now returns the total number of tokens used, instead of the list of all usage of the tokens, this simplifies the `get_submission_before_due_date()`.
- Updated the following function definition to avoid unnecessary calls to database:
    - Old:
        - `get_max_user_tokens(course_id, username)`
        - `get_user_token_usage(course_id, username)`
    - New:
        - `get_max_user_tokens(course_id, user, group_id)`
        - `get_user_token_usage(course_id, username, task)`


## Test Guidelines

1. In `/backend/` run the following and ensure that they all passes:
 ```bash 
npm test submission_student_check
```
<img width="710" alt="Screenshot 2024-04-25 at 2 54 09 PM" src="https://github.com/Kianoosh76/IBS/assets/71187784/331bfa49-464e-427c-a13d-794fb621facd">


## Checklist

### For PR author:
- [ ] I have added appropriate comments to the code.
- [ ] I have ensured code style and formatting are consistent.
- [ ] I have included unit tests or integration tests where applicable.
- [ ] I have considered edge cases and handled them appropriately.
- [ ] I have celebrated my accomplishments 🎉

### For PR reviewer:
- [ ] I have reviewed the code changes.
- [ ] I have verified that comments are clear and helpful.
- [ ] I have ensured code style and formatting align with project standards.
- [ ] I have examined any included tests and their coverage.
- [ ] I have considered potential edge cases and provided feedback on them.
